File: java/src/main/java/com/pqc/slhdsa/SLHDSA.java
package com.pqc.slhdsa;

import java.security.SecureRandom;
import java.util.Arrays;

/**
 * SLH-DSA (Stateless Hash-Based Digital Signature Algorithm) per FIPS 205.
 * Public API: keyGen, sign, verify.
 */
public final class SLHDSA {

    private SLHDSA() {}

    /**
     * Key pair: (secretKey, publicKey).
     */
    public record KeyPair(byte[] secretKey, byte[] publicKey) {}

    /**
     * slh_keygen(params): Generate an SLH-DSA key pair
     * (Algorithm 17 in FIPS 205).
     */
    public static KeyPair keyGen(SlhParams params) {
        return keyGen(params, new SecureRandom());
    }

    /**
     * slh_keygen with explicit random source (for testing).
     */
    public static KeyPair keyGen(SlhParams params, SecureRandom random) {
        int n = params.n;
        SlhHash hash = SlhHash.create(params);

        byte[] skSeed = new byte[n];
        byte[] skPrf = new byte[n];
        byte[] pkSeed = new byte[n];
        random.nextBytes(skSeed);
        random.nextBytes(skPrf);
        random.nextBytes(pkSeed);

        Address adrs = new Address();
        adrs.setLayerAddress(params.d - 1);
        adrs.setTreeAddress(0);

        byte[] pkRoot = Xmss.node(skSeed, 0, params.hPrime, pkSeed, adrs, params, hash);

        byte[] sk = SlhUtils.concat(skSeed, skPrf, pkSeed, pkRoot);
        byte[] pk = SlhUtils.concat(pkSeed, pkRoot);

        return new KeyPair(sk, pk);
    }

    /**
     * slh_sign(M, SK, params): Generate an SLH-DSA signature
     * (Algorithm 18 in FIPS 205, hedged variant).
     */
    public static byte[] sign(byte[] msg, byte[] sk, SlhParams params) {
        return sign(msg, sk, params, new SecureRandom());
    }

    /**
     * slh_sign with explicit random source.
     */
    public static byte[] sign(byte[] msg, byte[] sk, SlhParams params, SecureRandom random) {
        int n = params.n;
        SlhHash hash = SlhHash.create(params);

        // Parse secret key
        byte[] skSeed = SlhUtils.slice(sk, 0, n);
        byte[] skPrf  = SlhUtils.slice(sk, n, n);
        byte[] pkSeed = SlhUtils.slice(sk, 2 * n, n);
        byte[] pkRoot = SlhUtils.slice(sk, 3 * n, n);

        // Generate randomizer (hedged signing)
        byte[] optRand = new byte[n];
        random.nextBytes(optRand);

        // Generate randomized message digest
        byte[] r = hash.prfMsg(skPrf, optRand, msg, n);

        // Compute message digest
        byte[] digest = hash.hMsg(r, pkSeed, pkRoot, msg, params.m);

        // Split digest into md (FORS message) and tree/leaf indices
        // md = first floor(k*a / 8) bytes
        int mdBytes = (params.k * params.a + 7) / 8;
        byte[] md = SlhUtils.slice(digest, 0, mdBytes);

        // Compute tree index and leaf index from remaining digest bytes
        int treeBits = params.h - params.hPrime;
        int leafBits = params.hPrime;
        int treeBytes = (treeBits + 7) / 8;
        int leafBytes = (leafBits + 7) / 8;

        byte[] treeIdxBytes = SlhUtils.slice(digest, mdBytes, treeBytes);
        byte[] leafIdxBytes = SlhUtils.slice(digest, mdBytes + treeBytes, leafBytes);

        long idxTree = SlhUtils.toLong(treeIdxBytes, 0, treeBytes);
        idxTree &= ((1L << treeBits) - 1); // mask to treeBits
        int idxLeaf = SlhUtils.toInt(leafIdxBytes, 0, leafBytes);
        idxLeaf &= ((1 << leafBits) - 1); // mask to leafBits

        // Generate FORS signature
        Address adrs = new Address();
        adrs.setLayerAddress(0);
        adrs.setTreeAddress(idxTree);
        adrs.setType(Address.FORS_TREE);
        adrs.setKeyPairAddress(idxLeaf);

        byte[] sigFors = Fors.sign(md, skSeed, pkSeed, adrs, params, hash);

        // Get FORS public key for hypertree signing
        byte[] pkFors = Fors.pkFromSig(sigFors, md, pkSeed, adrs, params, hash);

        // Generate hypertree signature
        byte[] sigHt = Hypertree.sign(pkFors, skSeed, pkSeed, idxTree, idxLeaf, params, hash);

        // SIG = R || SIG_FORS || SIG_HT
        return SlhUtils.concat(r, sigFors, sigHt);
    }

    /**
     * slh_verify(M, SIG, PK, params): Verify an SLH-DSA signature
     * (Algorithm 19 in FIPS 205).
     */
    public static boolean verify(byte[] msg, byte[] sig, byte[] pk, SlhParams params) {
        int n = params.n;
        SlhHash hash = SlhHash.create(params);

        // Check signature length
        if (sig.length != params.sigSize()) {
            return false;
        }

        // Parse public key
        byte[] pkSeed = SlhUtils.slice(pk, 0, n);
        byte[] pkRoot = SlhUtils.slice(pk, n, n);

        // Parse signature: R || SIG_FORS || SIG_HT
        int offset = 0;
        byte[] r = SlhUtils.slice(sig, offset, n);
        offset += n;

        int forsSigSize = params.k * (params.a + 1) * n;
        byte[] sigFors = SlhUtils.slice(sig, offset, forsSigSize);
        offset += forsSigSize;

        byte[] sigHt = SlhUtils.slice(sig, offset, sig.length - offset);

        // Compute message digest
        byte[] digest = hash.hMsg(r, pkSeed, pkRoot, msg, params.m);

        // Split digest
        int mdBytes = (params.k * params.a + 7) / 8;
        byte[] md = SlhUtils.slice(digest, 0, mdBytes);

        int treeBits = params.h - params.hPrime;
        int leafBits = params.hPrime;
        int treeBytes = (treeBits + 7) / 8;
        int leafBytes = (leafBits + 7) / 8;

        byte[] treeIdxBytes = SlhUtils.slice(digest, mdBytes, treeBytes);
        byte[] leafIdxBytes = SlhUtils.slice(digest, mdBytes + treeBytes, leafBytes);

        long idxTree = SlhUtils.toLong(treeIdxBytes, 0, treeBytes);
        idxTree &= ((1L << treeBits) - 1);
        int idxLeaf = SlhUtils.toInt(leafIdxBytes, 0, leafBytes);
        idxLeaf &= ((1 << leafBits) - 1);

        // Compute FORS public key from signature
        Address adrs = new Address();
        adrs.setLayerAddress(0);
        adrs.setTreeAddress(idxTree);
        adrs.setType(Address.FORS_TREE);
        adrs.setKeyPairAddress(idxLeaf);

        byte[] pkFors = Fors.pkFromSig(sigFors, md, pkSeed, adrs, params, hash);

        // Verify hypertree signature
        return Hypertree.verify(pkFors, sigHt, pkSeed, idxTree, idxLeaf, pkRoot, params, hash);
    }
}

File: dotnet/src/PqcStandards/SlhDsa/SlhDsa.cs
using System.Security.Cryptography;

namespace PqcStandards.SlhDsa;

/// <summary>SLH-DSA digital signature algorithm (FIPS 205).</summary>
public static class SlhDsaAlgorithm
{
    /// <summary>Create the hash suite for the given params.</summary>
    public static IHashSuite GetHashSuite(SlhParams p) =>
        p.HashFamily == "SHAKE" ? new ShakeHash() : new Sha2Hash();

    /// <summary>SLH-DSA.KeyGen: returns (pk, sk).</summary>
    public static (byte[] pk, byte[] sk) KeyGen(SlhParams p)
    {
        byte[] skSeed = RandomNumberGenerator.GetBytes(p.N);
        byte[] skPrf = RandomNumberGenerator.GetBytes(p.N);
        byte[] pkSeed = RandomNumberGenerator.GetBytes(p.N);
        return KeyGenInternal(p, skSeed, skPrf, pkSeed);
    }

    internal static (byte[] pk, byte[] sk) KeyGenInternal(SlhParams p, byte[] skSeed, byte[] skPrf, byte[] pkSeed)
    {
        var hash = GetHashSuite(p);
        Address adrs = new();
        adrs.SetLayerAddress(p.D - 1);
        byte[] pkRoot = Xmss.Node(hash, pkSeed, skSeed, 0, p.HPrime, adrs, p);

        // pk = (pkSeed || pkRoot)
        byte[] pk = new byte[2 * p.N];
        Buffer.BlockCopy(pkSeed, 0, pk, 0, p.N);
        Buffer.BlockCopy(pkRoot, 0, pk, p.N, p.N);

        // sk = (skSeed || skPrf || pkSeed || pkRoot)
        byte[] sk = new byte[4 * p.N];
        Buffer.BlockCopy(skSeed, 0, sk, 0, p.N);
        Buffer.BlockCopy(skPrf, 0, sk, p.N, p.N);
        Buffer.BlockCopy(pkSeed, 0, sk, 2 * p.N, p.N);
        Buffer.BlockCopy(pkRoot, 0, sk, 3 * p.N, p.N);

        return (pk, sk);
    }

    /// <summary>SLH-DSA.Sign: returns signature.</summary>
    public static byte[] Sign(SlhParams p, byte[] sk, byte[] message)
    {
        int n = p.N;
        byte[] skSeed = sk[..n];
        byte[] skPrf = sk[n..(2 * n)];
        byte[] pkSeed = sk[(2 * n)..(3 * n)];
        byte[] pkRoot = sk[(3 * n)..];

        var hash = GetHashSuite(p);

        // Randomized: opt_rand = random N bytes
        byte[] optRand = RandomNumberGenerator.GetBytes(n);

        // R = PRF_msg(sk_prf, opt_rand, M)
        byte[] R = hash.PrfMsg(skPrf, optRand, message, n);

        // digest = H_msg(R, pk_seed, pk_root, M)
        int mdLen = (p.K * p.A + 7) / 8;  // FORS message length
        int treeBytes = (p.H - p.HPrime + 7) / 8;
        int leafBytes = (p.HPrime + 7) / 8;
        int digestLen = mdLen + treeBytes + leafBytes;
        byte[] digest = hash.Hmsg(R, pkSeed, pkRoot, message, digestLen);

        byte[] md = digest[..mdLen];
        long treeMask = (p.H - p.HPrime >= 64) ? -1L : (1L << (p.H - p.HPrime)) - 1;
        long idxTree = BytesToLong(digest, mdLen, treeBytes) & treeMask;
        int idxLeaf = (int)(BytesToLong(digest, mdLen + treeBytes, leafBytes) & ((1L << p.HPrime) - 1));

        // FORS sign
        Address adrs = new();
        adrs.SetLayerAddress(0);
        adrs.SetTreeAddress(idxTree);
        adrs.SetType(Address.ForsTree);
        adrs.SetKeyPairAddress(idxLeaf);

        var (forsSks, forsAuthPaths) = Fors.Sign(hash, pkSeed, skSeed, md, adrs, p);
        byte[] forsPk = Fors.PkFromSig(hash, pkSeed, forsSks, forsAuthPaths, md, adrs, p);

        // Hypertree sign on FORS public key
        byte[] htSig = Hypertree.Sign(hash, pkSeed, skSeed, forsPk, idxTree, idxLeaf, p);

        // Assemble signature: R || FORS sig (K trees × (1 sk + A auth nodes)) || HT sig
        using var sigStream = new MemoryStream();
        sigStream.Write(R);
        for (int i = 0; i < p.K; i++)  // K = number of FORS trees
        {
            sigStream.Write(forsSks[i]);
            for (int j = 0; j < p.A; j++)  // A = tree height = auth path length
                sigStream.Write(forsAuthPaths[i][j]);
        }
        sigStream.Write(htSig);

        return sigStream.ToArray();
    }

    /// <summary>SLH-DSA.Verify: returns true if signature is valid.</summary>
    public static bool Verify(SlhParams p, byte[] pk, byte[] message, byte[] sig)
    {
        int n = p.N;
        byte[] pkSeed = pk[..n];
        byte[] pkRoot = pk[n..];

        var hash = GetHashSuite(p);

        int offset = 0;
        byte[] R = sig[offset..(offset + n)];
        offset += n;

        // Recompute digest
        int mdLen = (p.K * p.A + 7) / 8;
        int treeBytes = (p.H - p.HPrime + 7) / 8;
        int leafBytes = (p.HPrime + 7) / 8;
        int digestLen = mdLen + treeBytes + leafBytes;
        byte[] digest = hash.Hmsg(R, pkSeed, pkRoot, message, digestLen);

        byte[] md = digest[..mdLen];
        long treeMask2 = (p.H - p.HPrime >= 64) ? -1L : (1L << (p.H - p.HPrime)) - 1;
        long idxTree = BytesToLong(digest, mdLen, treeBytes) & treeMask2;
        int idxLeaf = (int)(BytesToLong(digest, mdLen + treeBytes, leafBytes) & ((1L << p.HPrime) - 1));

        // Parse FORS signature: K trees × (1 sk + A auth nodes)
        byte[][] forsSks = new byte[p.K][];
        byte[][][] forsAuthPaths = new byte[p.K][][];
        for (int i = 0; i < p.K; i++)  // K = number of FORS trees
        {
            forsSks[i] = sig[offset..(offset + n)];
            offset += n;
            forsAuthPaths[i] = new byte[p.A][];
            for (int j = 0; j < p.A; j++)  // A = tree height = auth path length
            {
                forsAuthPaths[i][j] = sig[offset..(offset + n)];
                offset += n;
            }
        }

        // Recover FORS public key
        Address adrs = new();
        adrs.SetLayerAddress(0);
        adrs.SetTreeAddress(idxTree);
        adrs.SetType(Address.ForsTree);
        adrs.SetKeyPairAddress(idxLeaf);
        byte[] forsPk = Fors.PkFromSig(hash, pkSeed, forsSks, forsAuthPaths, md, adrs, p);

        // Verify hypertree signature
        byte[] htSig = sig[offset..];
        return Hypertree.Verify(hash, pkSeed, forsPk, htSig, idxTree, idxLeaf, pkRoot, p);
    }

    private static long BytesToLong(byte[] buf, int offset, int len)
    {
        long val = 0;
        for (int i = 0; i < len && (offset + i) < buf.Length; i++)
            val = (val << 8) | buf[offset + i];
        return val;
    }
}
